rates_api/utils/db.py

```diff
@@ -15,62 +14,47 @@ def connect_db() -> Any:
     return conn
 
 
-def get_available_dates(
-    date_from: str, date_to: str, orig: str, dest: str
-) -> Tuple[datetime.date, datetime.date] | None:
-    date_from_f = convert_date_type(date_from)
-    date_to_f = convert_date_type(date_to)
-    conn = connect_db()
-    with conn.cursor() as cur:
-        cur.execute(
-            "SELECT DISTINCT day "
-            "FROM prices "
-            "WHERE orig_code=%(org)s AND dest_code=%(dst)s "
-            "ORDER BY day;",
-            {
-                "org": orig.upper(),
-                "dst": dest.upper(),
-            },
-        )
-        unique_dates = cur.fetchall()
-        conn.close()
-        if not unique_dates:
-            return None
-    actual_dates = []
-    for date in unique_dates:
-        actual_dates.append(date)
-    start = actual_dates[0][0]
-    end = actual_dates[-1][0]
-    if (date_from_f < start) and (date_to_f > end):
-        return start, end
-    elif date_from_f < start < date_to_f <= end:
-        return start, date_to_f
-    elif (start <= date_from_f <= end) and (date_to_f > end):
-        return date_from_f, end
-    elif (date_from_f >= start) and (date_to_f <= end):
-        return date_from_f, date_to_f
-    return None
-
-
-def get_daily_rates(org: str, dst: str) -> List[Dict[str, Any]] | None:
+def get_daily_rates(
+    date_from: date, date_to: date, org: str, dst: str
+) -> List[Dict[str, Any]]:
+    """
+    Function selects data from database and calculates average prices
+    :param date_from:
+    :param date_to:
+    :param org:
+    :param dst:
+    :return: list of RealDicts included day and
+             average price (None if the conditions are not met)
+    """
     conn = connect_db()
     with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as dict_cr:
         dict_cr.execute(
-            "SELECT day, AVG(price) AS rate "
-            "FROM prices "
-            "WHERE orig_code=%(org)s AND dest_code=%(dst)s "
-            "GROUP BY day "
-            "HAVING COUNT(price) >= 3 "
-            "ORDER BY day;",
+            "WITH days AS ("
+            "SELECT generate_series("
+            "%(date_from)s::date, "
+            "%(date_to)s::date, "
+            "'1 day'::interval"
+            ")::date AS day) "
+            "SELECT days.day, CASE "
+            "WHEN COUNT(prices.price) >= 3 THEN AVG(prices.price)::INTEGER "
+            "ELSE NULL "
+            "END AS avg_price "
+            "FROM days "
+            "LEFT JOIN prices "
+            "ON days.day = prices.day "
+            "AND prices.orig_code = %(org)s "
+            "AND prices.dest_code = %(dst)s "
+            "GROUP BY days.day "
+            "ORDER BY days.day;",
```

Comment:
does it make this query easier to read?

```suggestion
            """
            WITH days AS (
                SELECT generate_series(
                    %(date_from)s::date,
                    %(date_to)s::date,
                    '1 day'::interval
                 )::date AS day
            )
            SELECT
                 days.day,
                 CASE
                     WHEN COUNT(prices.price) >= 3 
                     THEN AVG(prices.price)::INTEGER
                 END AS avg_price
            FROM days
            LEFT JOIN prices ON
                days.day = prices.day
                AND prices.orig_code = %(org)s
                AND prices.dest_code = %(dst)s
            GROUP BY days.day
            ORDER BY days.day;
            """,
```